cmake/OpenCVDetectCUDALanguage.cmake

```
@@ -0,0 +1,560 @@
+    set(CUDA_NVCC_FLAGS ${CUDA_NVCC_FLAGS} "-Xptxas;-dlcm=ca")
+  endif()
+
+  message(STATUS "CUDA NVCC target flags: ${CUDA_NVCC_FLAGS}")
+
+  OCV_OPTION(CUDA_FAST_MATH "Enable --use_fast_math for CUDA compiler " OFF)
+
+  if(CUDA_FAST_MATH)
+    set(CUDA_NVCC_FLAGS ${CUDA_NVCC_FLAGS} --use_fast_math)
+  endif()
+
+  OCV_OPTION(CUDA_ENABLE_DELAYLOAD "Enable delayed loading of CUDA DLLs" OFF VISIBLE_IF MSVC)
+
+  mark_as_advanced(CUDA_BUILD_CUBIN CUDA_BUILD_EMULATION CUDA_VERBOSE_BUILD CUDA_SDK_ROOT_DIR)
+
+  macro(ocv_cuda_filter_options)
+    foreach(var CMAKE_CXX_FLAGS CMAKE_CXX_FLAGS_RELEASE CMAKE_CXX_FLAGS_DEBUG)
+      set(${var}_backup_in_cuda_compile_ "${${var}}")
+
+      if (CV_CLANG)
+        # we remove -Winconsistent-missing-override and -Qunused-arguments
+        # just in case we are compiling CUDA with gcc but OpenCV with clang
+        string(REPLACE "-Winconsistent-missing-override" "" ${var} "${${var}}")
+        string(REPLACE "-Qunused-arguments" "" ${var} "${${var}}")
+      endif()
+
+      # we remove /EHa as it generates warnings under windows
+      string(REPLACE "/EHa" "" ${var} "${${var}}")
+
+      # we remove -ggdb3 flag as it leads to preprocessor errors when compiling CUDA files (CUDA 4.1)
+      string(REPLACE "-ggdb3" "" ${var} "${${var}}")
+
+      # we remove -Wsign-promo as it generates warnings under linux
+      string(REPLACE "-Wsign-promo" "" ${var} "${${var}}")
+
+      # we remove -Wno-sign-promo as it generates warnings under linux
+      string(REPLACE "-Wno-sign-promo" "" ${var} "${${var}}")
+
+      # we remove -Wno-delete-non-virtual-dtor because it's used for C++ compiler
+      # but NVCC uses C compiler by default
+      string(REPLACE "-Wno-delete-non-virtual-dtor" "" ${var} "${${var}}")
+
+      # we remove -frtti because it's used for C++ compiler
+      # but NVCC uses C compiler by default
+      string(REPLACE "-frtti" "" ${var} "${${var}}")
+
+      string(REPLACE "-fvisibility-inlines-hidden" "" ${var} "${${var}}")
+
+      # cc1: warning: command line option '-Wsuggest-override' is valid for C++/ObjC++ but not for C
+      string(REPLACE "-Wsuggest-override" "" ${var} "${${var}}")
+
+      # issue: #11552 (from OpenCVCompilerOptions.cmake)
+      string(REGEX REPLACE "-Wimplicit-fallthrough(=[0-9]+)? " "" ${var} "${${var}}")
+
+      # removal of custom specified options
+      if(OPENCV_CUDA_NVCC_FILTEROUT_OPTIONS)
+        foreach(__flag ${OPENCV_CUDA_NVCC_FILTEROUT_OPTIONS})
+          string(REPLACE "${__flag}" "" ${var} "${${var}}")
+        endforeach()
+      endif()
+    endforeach()
+  endmacro()
+
+  macro(ocv_cuda_unfilter_options)
+    foreach(var CMAKE_CXX_FLAGS CMAKE_CXX_FLAGS_RELEASE CMAKE_CXX_FLAGS_DEBUG)
+      set(${var} "${${var}_backup_in_cuda_compile_}")
+      unset(${var}_backup_in_cuda_compile_)
+    endforeach()
+  endmacro()
+
+  macro(ocv_nvcc_flags)
+    if(BUILD_SHARED_LIBS)
+      set(CUDA_NVCC_FLAGS ${CUDA_NVCC_FLAGS} -Xcompiler -DCVAPI_EXPORTS)
```

ccache is a little picky with `-Xcompiler` flag, this form will work with latest ccache

We should do the same modification on `cmake/OpenCVDetectCUDA.cmake`. 

Or should I do a separate PR, once this one is merged ?

```suggestion
      set(CUDA_NVCC_FLAGS ${CUDA_NVCC_FLAGS} -Xcompiler=-DCVAPI_EXPORTS)
```

---

> ccache is a little picky with -Xcompiler flag, this form will work with latest ccache

I don't use ccache (at the moment) how can I tell if these fixes worked as it builds fine either way?

> We should do the same modification on cmake/OpenCVDetectCUDA.cmake.

I'm not sure if its necessary if ccache doesn't work when CUDA is not a first class language?

---

You could subjectively tell by the speed of recompile:
````
make
make clean
make
````

Second make should be faster fall ALL files (not only CXX/C)

For a more objective/quantitative way, you need to see if compiling CUDA is cached and/or if is cached on first compile.
You will need to add log to ccache in `$HOME/.ccache/ccache.conf` a line
```
log_file = /home/dgeld/.ccache/log.txt
```

In the log, you will see that the caching is failing : `Failed; falling back to running the real compiler`

```
[2023-08-26T08:38:00.211404 10953] Hostname: dgeld-VirtualBox
[2023-08-26T08:38:00.211407 10953] Working directory: /home/dgeld/src/david/opencv/opencv_cuda/tmp2/opencv/build/modules/core/CMakeFiles/cuda_compile_1.dir/src/cuda
[2023-08-26T08:38:00.211414 10953] Compiler: /usr/local/cuda-11.4/bin/nvcc
[2023-08-26T08:38:00.211417 10953] Compiler type: nvcc
[2023-08-26T08:38:00.211460 10953] ,"-fdata-sections","-Wa,--noexecstack","-fsigned-char","-Wno-psabi","-fsigned-char","-W","-Wall","-Wreturn-type","-Wnon-virtual-dtor","-Waddress","-Wsequence-point","-Wformat","-Wformat-security","-Wmissing-declarations","-Winit-self","-Wpointer-arith","-Wuninitialized","-Wno-comment","-Wno-strict-overflow","-fdiagnostics-show-option","-pthread","-fomit-frame-pointer","-ffunction-sections","-fdata-sections","-fvisibility=hidden","-Wno-undef","-Wno-enum-compare","-Wno-unused-function","-Wno-shadow","-Wno-unused-but-set-variable","-O3","-DNDEBUG","-DNDEBUG" is not a regular file, not considering as input file
[2023-08-26T08:38:00.211468 10953] arch=compute_72,code=sm_72 is not a regular file, not considering as input file
[2023-08-26T08:38:00.211474 10953] arch=compute_87,code=sm_87 is not a regular file, not considering as input file
[2023-08-26T08:38:00.211523 10953] Source file: /home/dgeld/src/david/opencv/opencv_cuda/tmp2/opencv/modules/core/src/cuda/gpu_mat_nd.cu
[2023-08-26T08:38:00.211527 10953] Object file: /home/dgeld/src/david/opencv/opencv_cuda/tmp2/opencv/build/modules/core/CMakeFiles/cuda_compile_1.dir/src/cuda/./cuda_compile_1_generated_gpu_mat_nd.cu.o
[2023-08-26T08:38:00.211588 10953] Hashing CWD /home/dgeld/src/david/opencv/opencv_cuda/tmp2/opencv/build/modules/core/CMakeFiles/cuda_compile_1.dir/src/cuda
[2023-08-26T08:38:00.211597 10953] Trying direct lookup
[2023-08-26T08:38:00.211660 10953] Filesystem has less than 100 MiB free space, not using inode cache
[2023-08-26T08:38:00.211668 10953] Manifest key: ef40rbo157eoo2lm5l0alo1p57ehhujkk
[2023-08-26T08:38:00.211676 10953] No ef40rbo157eoo2lm5l0alo1p57ehhujkk in local storage
[2023-08-26T08:38:00.211775 10953] Running preprocessor
[2023-08-26T08:38:00.211797 10953] Executing /usr/local/cuda-11.4/bin/nvcc -m64 -Xcompiler ,"-fdata-sections","-Wa,--noexecstack","-fsigned-char","-Wno-psabi","-fsigned-char","-W","-Wall","-Wreturn-type","-Wnon-virtual-dtor","-Waddress","-Wsequence-point","-Wformat","-Wformat-security","-Wmissing-declarations","-Winit-self","-Wpointer-arith","-Wuninitialized","-Wno-comment","-Wno-strict-overflow","-fdiagnostics-show-option","-pthread","-fomit-frame-pointer","-ffunction-sections","-fdata-sections","-fvisibility=hidden","-Wno-undef","-Wno-enum-compare","-Wno-unused-function","-Wno-shadow","-Wno-unused-but-set-variable","-O3","-DNDEBUG","-DNDEBUG" -gencode arch=compute_72,code=sm_72 -gencode arch=compute_87,code=sm_87 -Xcompiler -Xcompiler -fPIC --std=c++14 -ccbin /usr/bin/aarch64-linux-gnu-gcc -D_USE_MATH_DEFINES -D__STDC_CONSTANT_MACROS -D__STDC_LIMIT_MACROS -D__STDC_FORMAT_MACROS -DOPENCV_ALLOCATOR_STATS_COUNTER_TYPE=int -D__OPENCV_BUILD=1 -D_FORCE_INLINES -DCVAPI_EXPORTS -DNVCC -I/usr/local/cuda-11.4/targets/aarch64-linux/include -I/home/dgeld/src/david/opencv/opencv_cuda/tmp2/opencv/build -I/home/dgeld/src/david/opencv/opencv_cuda/tmp2/opencv/modules/core/include -I/home/dgeld/src/david/opencv/opencv_cuda/tmp2/opencv/build/modules/core -I/home/dgeld/src/david/opencv/opencv_cuda/tmp2/opencv_contrib/modules/cudev/include -I/home/dgeld/src/david/opencv/opencv_cuda/tmp2/opencv/3rdparty/zlib -I/home/dgeld/src/david/opencv/opencv_cuda/tmp2/opencv/build/3rdparty/zlib -E -o /run/user/1000/ccache-tmp/cpp_stdout.tmp.80bAgr.i /home/dgeld/src/david/opencv/opencv_cuda/tmp2/opencv/modules/core/src/cuda/gpu_mat_nd.cu
[2023-08-26T08:38:00.213663 10953] Preprocessor gave exit status 1
[2023-08-26T08:38:00.213684 10953] Failed; falling back to running the real compiler
[2023-08-26T08:38:00.213696 10953] Executing /usr/local/cuda-11.4/bin/nvcc /home/dgeld/src/david/opencv/opencv_cuda/tmp2/opencv/modules/core/src/cuda/gpu_mat_nd.cu -c -o /home/dgeld/src/david/opencv/opencv_cuda/tmp2/opencv/build/modules/core/CMakeFiles/cuda_compile_1.dir/src/cuda/./cuda_compile_1_generated_gpu_mat_nd.cu.o -ccbin /usr/bin/aarch64-linux-gnu-gcc -m64 -D_USE_MATH_DEFINES -D__STDC_CONSTANT_MACROS -D__STDC_LIMIT_MACROS -D__STDC_FORMAT_MACROS -DOPENCV_ALLOCATOR_STATS_COUNTER_TYPE=int -D__OPENCV_BUILD=1 -Xcompiler ,"-fdata-sections","-Wa,--noexecstack","-fsigned-char","-Wno-psabi","-fsigned-char","-W","-Wall","-Wreturn-type","-Wnon-virtual-dtor","-Waddress","-Wsequence-point","-Wformat","-Wformat-security","-Wmissing-declarations","-Winit-self","-Wpointer-arith","-Wuninitialized","-Wno-comment","-Wno-strict-overflow","-fdiagnostics-show-option","-pthread","-fomit-frame-pointer","-ffunction-sections","-fdata-sections","-fvisibility=hidden","-Wno-undef","-Wno-enum-compare","-Wno-unused-function","-Wno-shadow","-Wno-unused-but-set-variable","-O3","-DNDEBUG","-DNDEBUG" -gencode arch=compute_72,code=sm_72 -gencode arch=compute_87,code=sm_87 -D_FORCE_INLINES -Xcompiler -DCVAPI_EXPORTS -Xcompiler -fPIC --std=c++14 -Xlinker --unresolved-symbols=ignore-in-shared-libs -DNVCC -I/usr/local/cuda-11.4/targets/aarch64-linux/include -I/home/dgeld/src/david/opencv/opencv_cuda/tmp2/opencv/build -I/home/dgeld/src/david/opencv/opencv_cuda/tmp2/opencv/modules/core/include -I/home/dgeld/src/david/opencv/opencv_cuda/tmp2/opencv/build/modules/core -I/home/dgeld/src/david/opencv/opencv_cuda/tmp2/opencv_contrib/modules/cudev/include -I/home/dgeld/src/david/opencv/opencv_cuda/tmp2/opencv/3rdparty/zlib -I/home/dgeld/src/david/opencv/opencv_cuda/tmp2/opencv/build/3rdparty/zlib
[2023-08-26T08:38:00.213700 10953] Result: direct_cache_miss
[2023-08-26T08:38:00.213703 10953] Result: local_storage_read_miss
[2023-08-26T08:38:00.213705 10953] Result: preprocessor_error
```

when it works, you will see, ccache storing:

```

[2023-08-26T08:49:41.990771 5558 ] Storing embedded entry #1 .o (3344152 bytes) from /home/dgeld/src/david/opencv/opencv_cuda/tmp2/opencv/build/modules/core/CMakeFiles/cuda_compile_1.dir/src/cuda/./cuda_compile_1_generated_gpu_mat.cu.o
```
or a cache hit

```
[2023-08-26T08:52:36.495780 8494 ] Compiler: /usr/local/cuda-11.4/bin/nvcc
[2023-08-26T08:52:36.495782 8494 ] Compiler type: nvcc
[2023-08-26T08:52:36.495827 8494 ] ,"-fdata-sections","-Wa,--noexecstack","-fsigned-char","-Wno-psabi","-fsigned-char","-W","-Wall","-Wreturn-type","-Wnon-virtual-dtor","-Waddress","-Wsequence-point","-Wformat","-Wformat-security","-Wmissing-declarations","-Winit-self","-Wpointer-arith","-Wuninitialized","-Wno-comment","-Wno-strict-overflow","-fdiagnostics-show-option","-pthread","-fomit-frame-pointer","-ffunction-sections","-fdata-sections","-fvisibility=hidden","-Wno-undef","-Wno-enum-compare","-Wno-unused-function","-Wno-shadow","-Wno-unused-but-set-variable","-O3","-DNDEBUG","-DNDEBUG" is not a regular file, not considering as input file
[2023-08-26T08:52:36.495833 8494 ] arch=compute_72,code=sm_72 is not a regular file, not considering as input file
[2023-08-26T08:52:36.495841 8494 ] arch=compute_87,code=sm_87 is not a regular file, not considering as input file
[2023-08-26T08:52:36.495886 8494 ] Source file: /home/dgeld/src/david/opencv/opencv_cuda/tmp2/opencv/modules/core/src/cuda/gpu_mat.cu
[2023-08-26T08:52:36.495890 8494 ] Object file: /home/dgeld/src/david/opencv/opencv_cuda/tmp2/opencv/build/modules/core/CMakeFiles/cuda_compile_1.dir/src/cuda/./cuda_compile_1_generated_gpu_mat.cu.o
[2023-08-26T08:52:36.495951 8494 ] Hashing CWD /home/dgeld/src/david/opencv/opencv_cuda/tmp2/opencv/build/modules/core/CMakeFiles/cuda_compile_1.dir/src/cuda
[2023-08-26T08:52:36.495960 8494 ] Trying direct lookup
[2023-08-26T08:52:36.496036 8494 ] Filesystem has less than 100 MiB free space, not using inode cache
[2023-08-26T08:52:36.496048 8494 ] Manifest key: 02cer5pejds1qpa3bqlfbgj8f8s31ktbk
[2023-08-26T08:52:36.496065 8494 ] Retrieved 02cer5pejds1qpa3bqlfbgj8f8s31ktbk from local storage (/home/dgeld/.ccache/0/2/cer5pejds1qpa3bqlfbgj8f8s31ktbkM)
[2023-08-26T08:52:36.498064 8494 ] Got result key from manifest
[2023-08-26T08:52:36.498073 8494 ] Result key: 00a48hbqsck6d582k5lv8svag1vj5t9q6
[2023-08-26T08:52:36.498327 8494 ] Retrieved 00a48hbqsck6d582k5lv8svag1vj5t9q6 from local storage (/home/dgeld/.ccache/0/0/a48hbqsck6d582k5lv8svag1vj5t9q6R)
[2023-08-26T08:52:36.500838 8494 ] Reading embedded entry #0 <stderr> (614 bytes)
[2023-08-26T08:52:36.500877 8494 ] Reading embedded entry #1 .o (3344152 bytes)
[2023-08-26T08:52:36.500891 8494 ] Writing to /home/dgeld/src/david/opencv/opencv_cuda/tmp2/opencv/build/modules/core/CMakeFiles/cuda_compile_1.dir/src/cuda/./cuda_compile_1_generated_gpu_mat.cu.o
[2023-08-26T08:52:36.502660 8494 ] Succeeded getting cached result
[2023-08-26T08:52:36.502688 8494 ] Result: direct_cache_hit
[2023-08-26T08:52:36.502693 8494 ] Result: local_storage_hit
[2023-08-26T08:52:36.502696 8494 ] Result: local_storage_read_hit
[2023-08-26T08:52:36.502698 8494 ] Result: local_storage_read_hit
```

---

> > We should do the same modification on cmake/OpenCVDetectCUDA.cmake.
> 
> I'm not sure if its necessary if ccache doesn't work when CUDA is not a first class language?

I will submit a PR to CMake so their `run_nvcc.cmake` takes into account ccache.

We could postpone this change to later but it could be better to sync the changes as it does not break the current build.

---

> > ccache is a little picky with -Xcompiler flag, this form will work with latest ccache
> 
> I don't use ccache (at the moment) how can I tell if these fixes worked as it builds fine either way?

You need to have `ccache` installed and OpenCV should find it when using `-DENABLE_CCACHE=ON`

---

Where can I observe the errors realted to the missing "=" sign in the log?  My output from without these changes for **mean_shift.cu** is

> [2023-08-28T17:36:27.871128 7577 ] === CCACHE 4.8.2 STARTED =========================================
> [2023-08-28T17:36:27.871138 7577 ] Configuration file: /home/b/.ccache/ccache.conf
> [2023-08-28T17:36:27.871140 7577 ] System configuration file: /usr/local/etc/ccache.conf
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) absolute_paths_in_stderr = false
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) base_dir =
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) cache_dir = /home/b/.ccache
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) compiler =
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) compiler_check = mtime
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) compiler_type = auto
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) compression = true
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) compression_level = 0
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) cpp_extension =
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) debug = false
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) debug_dir =
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) depend_mode = false
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) direct_mode = true
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) disable = false
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) extra_files_to_hash =
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) file_clone = false
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) hard_link = false
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) hash_dir = true
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) ignore_headers_in_manifest =
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) ignore_options =
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) inode_cache = true
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) keep_comments_cpp = false
> [2023-08-28T17:36:27.871140 7577 ] Config: (/home/b/.ccache/ccache.conf) log_file = /home/b/.ccache/log.txt
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) max_files = 0
> [2023-08-28T17:36:27.871140 7577 ] Config: (/home/b/.ccache/ccache.conf) max_size = 5.0 GB
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) msvc_dep_prefix = Note: including file:
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) namespace =
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) path =
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) pch_external_checksum = false
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) prefix_command =
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) prefix_command_cpp =
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) read_only = false
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) read_only_direct = false
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) recache = false
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) remote_only = false
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) remote_storage =
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) reshare = false
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) run_second_cpp = true
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) sloppiness =
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) stats = true
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) stats_log =
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) temporary_dir = /mnt/wslg/runtime-dir/ccache-tmp
> [2023-08-28T17:36:27.871140 7577 ] Config: (default) umask =
> [2023-08-28T17:36:27.871160 7577 ] Command line: /usr/local/bin/ccache /usr/local/cuda/bin/nvcc -forward-unknown-to-host-compiler -DCVAPI_EXPORTS -D_USE_MATH_DEFINES -D__OPENCV_BUILD=1 -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/home/b/build/opencv/cudaimgproc_ccache_log/3rdparty/ippicv/ippicv_lnx/icv/include -I/home/b/build/opencv/cudaimgproc_ccache_log/3rdparty/ippicv/ippicv_lnx/iw/include -I/home/b/repos/opencv/opencv_contrib/modules/cudaimgproc/include -I/home/b/build/opencv/cudaimgproc_ccache_log/modules/cudaimgproc -I/home/b/repos/opencv/opencv_contrib/modules/cudev/include -I/home/b/repos/opencv/opencv/modules/core/include -I/home/b/repos/opencv/opencv/modules/imgproc/include -isystem=/home/b/build/opencv/cudaimgproc_ccache_log -isystem=/usr/local/cuda/include -O3 -DNDEBUG --generate-code=arch=compute_86,code=[sm_86] -Xcompiler=-fPIC -D_FORCE_INLINES -Xcompiler -DCVAPI_EXPORTS -fPIC "-Xcompiler=   -fsigned-char -W -Wall -Wreturn-type -Wnon-virtual-dtor -Waddress -Wsequence-point -Wformat -Wformat-security -Winit-self -Wpointer-arith  -Wuninitialized   -Wno-comment -Wno-strict-overflow -fdiagnostics-show-option -Wno-long-long -pthread -fomit-frame-pointer -ffunction-sections -fdata-sections  -msse -msse2 -msse3 -fvisibility=hidden  -Wno-undef -Wno-missing-declarations -Wno-shadow -Wno-unused-parameter -Wno-unused-but-set-variable    -O3 -DNDEBUG  -DNDEBUG" -std=c++14 -MD -MT modules/cudaimgproc/CMakeFiles/opencv_cudaimgproc.dir/src/cuda/mean_shift.cu.o -MF CMakeFiles/opencv_cudaimgproc.dir/src/cuda/mean_shift.cu.o.d -x cu -c /home/b/repos/opencv/opencv_contrib/modules/cudaimgproc/src/cuda/mean_shift.cu -o CMakeFiles/opencv_cudaimgproc.dir/src/cuda/mean_shift.cu.o
> [2023-08-28T17:36:27.871166 7577 ] Hostname: DESKTOP-15JFLGE
> [2023-08-28T17:36:27.871167 7577 ] Working directory: /home/b/build/opencv/cudaimgproc_ccache_log/modules/cudaimgproc
> [2023-08-28T17:36:27.871172 7577 ] Compiler: /usr/local/cuda/bin/nvcc
> [2023-08-28T17:36:27.871173 7577 ] Compiler type: nvcc
> [2023-08-28T17:36:27.871276 7577 ] Source file: /home/b/repos/opencv/opencv_contrib/modules/cudaimgproc/src/cuda/mean_shift.cu
> [2023-08-28T17:36:27.871278 7577 ] Dependency file: CMakeFiles/opencv_cudaimgproc.dir/src/cuda/mean_shift.cu.o.d
> [2023-08-28T17:36:27.871280 7577 ] Object file: CMakeFiles/opencv_cudaimgproc.dir/src/cuda/mean_shift.cu.o
> [2023-08-28T17:36:27.871298 7577 ] Trying direct lookup
> [2023-08-28T17:36:27.871385 7577 ] Manifest key: 3f9dv79og5h165mh37qat5gddnnq1nhv0
> [2023-08-28T17:36:27.871396 7577 ] Retrieved 3f9dv79og5h165mh37qat5gddnnq1nhv0 from local storage (/home/b/.ccache/3/f/9dv79og5h165mh37qat5gddnnq1nhv0M)
> [2023-08-28T17:36:27.872096 7577 ] Got result key from manifest
> [2023-08-28T17:36:27.872101 7577 ] Result key: e5a3051eihi33fco691mmmm4lkvodeoiu
> [2023-08-28T17:36:27.872114 7577 ] Retrieved e5a3051eihi33fco691mmmm4lkvodeoiu from local storage (/home/b/.ccache/e/5/a3051eihi33fco691mmmm4lkvodeoiuR)
> [2023-08-28T17:36:27.872186 7577 ] Reading embedded entry #0 .o (52552 bytes)
> [2023-08-28T17:36:27.872188 7577 ] Writing to CMakeFiles/opencv_cudaimgproc.dir/src/cuda/mean_shift.cu.o
> [2023-08-28T17:36:27.872220 7577 ] Reading embedded entry #1 .d (18004 bytes)
> [2023-08-28T17:36:27.872221 7577 ] Writing to CMakeFiles/opencv_cudaimgproc.dir/src/cuda/mean_shift.cu.o.d
> [2023-08-28T17:36:27.872239 7577 ] Succeeded getting cached result
> [2023-08-28T17:36:27.872244 7577 ] Result: direct_cache_hit
> [2023-08-28T17:36:27.872245 7577 ] Result: local_storage_hit
> [2023-08-28T17:36:27.872246 7577 ] Result: local_storage_read_hit
> [2023-08-28T17:36:27.872247 7577 ] Result: local_storage_read_hit
> [2023-08-28T17:36:27.872249 7577 ] Acquiring /home/b/.ccache/9/9/stats.lock
> [2023-08-28T17:36:27.872267 7577 ] Acquired /home/b/.ccache/9/9/stats.lock
> [2023-08-28T17:36:27.872269 7577 ] Creating /home/b/.ccache/9/9/stats.alive
> [2023-08-28T17:36:27.872330 7577 ] Releasing /home/b/.ccache/9/9/stats.lock
> [2023-08-28T17:36:27.872337 7577 ] Unlink /home/b/.ccache/9/9/stats.alive
> [2023-08-28T17:36:27.872344 7577 ] Unlink /home/b/.ccache/9/9/stats.lock
> [2023-08-28T17:36:27.872345 7577 ] Released /home/b/.ccache/9/9/stats.lock
> [2023-08-28T17:36:27.872349 7577 ] Accumulated stats for inode cache: hits=0, misses=0, errors=0

and calling make after make clean takes a matter of seconds?

---

In your example, you pass the = sign so there is no error.

The 2 cases (not working and working) were in may initial comment. 

An example of a log with the error. It tries to run the preprocessor to get the dependencies and you get `Preprocessor gave exit status 1` then `Failed; falling back to running the real compiler`

```
[2023-08-28T20:10:19.192892 7814 ] Hostname: dgeld-VirtualBox
[2023-08-28T20:10:19.192895 7814 ] Working directory: /home/dgeld/src/david/opencv/opencv_cuda/tmp2/opencv/build/modules/core/CMakeFiles/cuda_compile_1.dir/src/cuda
[2023-08-28T20:10:19.192902 7814 ] Compiler: /usr/local/cuda-11.4/bin/nvcc
[2023-08-28T20:10:19.192905 7814 ] Compiler type: nvcc
[2023-08-28T20:10:19.192954 7814 ] ,"-fdata-sections","-Wa,--noexecstack","-fsigned-char","-Wno-psabi","-fsigned-char","-W","-Wall","-Wreturn-type","-Wnon-virtual-dtor","-Waddress","-Wsequence-point","-Wformat","-Wformat-security","-Wmissing-declarations","-Winit-self","-Wpointer-arith","-Wuninitialized","-Wno-comment","-Wno-strict-overflow","-fdiagnostics-show-option","-pthread","-fomit-frame-pointer","-ffunction-sections","-fdata-sections","-fvisibility=hidden","-Wno-undef","-Wno-enum-compare","-Wno-unused-function","-Wno-shadow","-Wno-unused-but-set-variable","-O3","-DNDEBUG","-DNDEBUG" is not a regular file, not considering as input file
[2023-08-28T20:10:19.192960 7814 ] arch=compute_72,code=sm_72 is not a regular file, not considering as input file
[2023-08-28T20:10:19.192964 7814 ] arch=compute_87,code=sm_87 is not a regular file, not considering as input file
[2023-08-28T20:10:19.192971 7814 ] arch=compute_87,code=compute_87 is not a regular file, not considering as input file
[2023-08-28T20:10:19.193018 7814 ] Source file: /home/dgeld/src/david/opencv/opencv_cuda/tmp2/opencv/modules/core/src/cuda/gpu_mat.cu
[2023-08-28T20:10:19.193021 7814 ] Object file: /home/dgeld/src/david/opencv/opencv_cuda/tmp2/opencv/build/modules/core/CMakeFiles/cuda_compile_1.dir/src/cuda/./cuda_compile_1_generated_gpu_mat.cu.o
[2023-08-28T20:10:19.193086 7814 ] Hashing CWD /home/dgeld/src/david/opencv/opencv_cuda/tmp2/opencv/build/modules/core/CMakeFiles/cuda_compile_1.dir/src/cuda
[2023-08-28T20:10:19.193095 7814 ] Trying direct lookup
[2023-08-28T20:10:19.193176 7814 ] Filesystem has less than 100 MiB free space, not using inode cache
[2023-08-28T20:10:19.193187 7814 ] Manifest key: 9a7bl7f9sraqg0t6ni9qqpg70njpef1fs
[2023-08-28T20:10:19.193197 7814 ] No 9a7bl7f9sraqg0t6ni9qqpg70njpef1fs in local storage
[2023-08-28T20:10:19.193326 7814 ] Running preprocessor
[2023-08-28T20:10:19.193350 7814 ] Executing /usr/local/cuda-11.4/bin/nvcc -m64 -Xcompiler ,"-fdata-sections","-Wa,--noexecstack","-fsigned-char","-Wno-psabi","-fsigned-char","-W","-Wall","-Wreturn-type","-Wnon-virtual-dtor","-Waddress","-Wsequence-point","-Wformat","-Wformat-security","-Wmissing-declarations","-Winit-self","-Wpointer-arith","-Wuninitialized","-Wno-comment","-Wno-strict-overflow","-fdiagnostics-show-option","-pthread","-fomit-frame-pointer","-ffunction-sections","-fdata-sections","-fvisibility=hidden","-Wno-undef","-Wno-enum-compare","-Wno-unused-function","-Wno-shadow","-Wno-unused-but-set-variable","-O3","-DNDEBUG","-DNDEBUG" -gencode arch=compute_72,code=sm_72 -gencode arch=compute_87,code=sm_87 -gencode arch=compute_87,code=compute_87 -Xcompiler -Xcompiler -fPIC --std=c++14 -ccbin /usr/bin/aarch64-linux-gnu-gcc -D_USE_MATH_DEFINES -D__STDC_CONSTANT_MACROS -D__STDC_LIMIT_MACROS -D__STDC_FORMAT_MACROS -DOPENCV_ALLOCATOR_STATS_COUNTER_TYPE=int -D__OPENCV_BUILD=1 -D_FORCE_INLINES -DCVAPI_EXPORTS -DNVCC -I/usr/local/cuda-11.4/targets/aarch64-linux/include -I/home/dgeld/src/david/opencv/opencv_cuda/tmp2/opencv/build -I/home/dgeld/src/david/opencv/opencv_cuda/tmp2/opencv/modules/core/include -I/home/dgeld/src/david/opencv/opencv_cuda/tmp2/opencv/build/modules/core -I/home/dgeld/src/david/opencv/opencv_cuda/tmp2/opencv_contrib/modules/cudev/include -I/home/dgeld/src/david/opencv/opencv_cuda/tmp2/opencv/3rdparty/zlib -I/home/dgeld/src/david/opencv/opencv_cuda/tmp2/opencv/build/3rdparty/zlib -E -o /run/user/1000/ccache-tmp/cpp_stdout.tmp.nVpzR1.i /home/dgeld/src/david/opencv/opencv_cuda/tmp2/opencv/modules/core/src/cuda/gpu_mat.cu
[2023-08-28T20:10:19.195320 7814 ] Preprocessor gave exit status 1
[2023-08-28T20:10:19.195339 7814 ] Failed; falling back to running the real compiler
```

---

> and calling make after make clean takes a matter of seconds?

Yes the compilation should be faster. Linking the libraries still need time.

---

That's the output without the changes
> -Xcompiler -DCVAPI_EXPORTS -fPIC 

with the changes I get the following

> [2023-08-28T21:55:15.968220 14016] === CCACHE 4.8.2 STARTED =========================================
> [2023-08-28T21:55:15.968242 14016] Configuration file: /home/b/.ccache/ccache.conf
> [2023-08-28T21:55:15.968245 14016] System configuration file: /usr/local/etc/ccache.conf
> [2023-08-28T21:55:15.968245 14016] Config: (default) absolute_paths_in_stderr = false
> [2023-08-28T21:55:15.968245 14016] Config: (default) base_dir =
> [2023-08-28T21:55:15.968245 14016] Config: (default) cache_dir = /home/b/.ccache
> [2023-08-28T21:55:15.968245 14016] Config: (default) compiler =
> [2023-08-28T21:55:15.968245 14016] Config: (default) compiler_check = mtime
> [2023-08-28T21:55:15.968245 14016] Config: (default) compiler_type = auto
> [2023-08-28T21:55:15.968245 14016] Config: (default) compression = true
> [2023-08-28T21:55:15.968245 14016] Config: (default) compression_level = 0
> [2023-08-28T21:55:15.968245 14016] Config: (default) cpp_extension =
> [2023-08-28T21:55:15.968245 14016] Config: (default) debug = false
> [2023-08-28T21:55:15.968245 14016] Config: (default) debug_dir =
> [2023-08-28T21:55:15.968245 14016] Config: (default) depend_mode = false
> [2023-08-28T21:55:15.968245 14016] Config: (default) direct_mode = true
> [2023-08-28T21:55:15.968245 14016] Config: (default) disable = false
> [2023-08-28T21:55:15.968245 14016] Config: (default) extra_files_to_hash =
> [2023-08-28T21:55:15.968245 14016] Config: (default) file_clone = false
> [2023-08-28T21:55:15.968245 14016] Config: (default) hard_link = false
> [2023-08-28T21:55:15.968245 14016] Config: (default) hash_dir = true
> [2023-08-28T21:55:15.968245 14016] Config: (default) ignore_headers_in_manifest =
> [2023-08-28T21:55:15.968245 14016] Config: (default) ignore_options =
> [2023-08-28T21:55:15.968245 14016] Config: (default) inode_cache = true
> [2023-08-28T21:55:15.968245 14016] Config: (default) keep_comments_cpp = false
> [2023-08-28T21:55:15.968245 14016] Config: (/home/b/.ccache/ccache.conf) log_file = /home/b/.ccache/log.txt
> [2023-08-28T21:55:15.968245 14016] Config: (default) max_files = 0
> [2023-08-28T21:55:15.968245 14016] Config: (/home/b/.ccache/ccache.conf) max_size = 5.0 GB
> [2023-08-28T21:55:15.968245 14016] Config: (default) msvc_dep_prefix = Note: including file:
> [2023-08-28T21:55:15.968245 14016] Config: (default) namespace =
> [2023-08-28T21:55:15.968245 14016] Config: (default) path =
> [2023-08-28T21:55:15.968245 14016] Config: (default) pch_external_checksum = false
> [2023-08-28T21:55:15.968245 14016] Config: (default) prefix_command =
> [2023-08-28T21:55:15.968245 14016] Config: (default) prefix_command_cpp =
> [2023-08-28T21:55:15.968245 14016] Config: (default) read_only = false
> [2023-08-28T21:55:15.968245 14016] Config: (default) read_only_direct = false
> [2023-08-28T21:55:15.968245 14016] Config: (default) recache = false
> [2023-08-28T21:55:15.968245 14016] Config: (default) remote_only = false
> [2023-08-28T21:55:15.968245 14016] Config: (default) remote_storage =
> [2023-08-28T21:55:15.968245 14016] Config: (default) reshare = false
> [2023-08-28T21:55:15.968245 14016] Config: (default) run_second_cpp = true
> [2023-08-28T21:55:15.968245 14016] Config: (default) sloppiness =
> [2023-08-28T21:55:15.968245 14016] Config: (default) stats = true
> [2023-08-28T21:55:15.968245 14016] Config: (default) stats_log =
> [2023-08-28T21:55:15.968245 14016] Config: (default) temporary_dir = /mnt/wslg/runtime-dir/ccache-tmp
> [2023-08-28T21:55:15.968245 14016] Config: (default) umask =
> [2023-08-28T21:55:15.968276 14016] Command line: /usr/local/bin/ccache /usr/local/cuda/bin/nvcc -forward-unknown-to-host-compiler -DCVAPI_EXPORTS -D_USE_MATH_DEFINES -D__OPENCV_BUILD=1 -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/home/b/build/opencv/cudaimgproc_ccache_log_no_equal/3rdparty/ippicv/ippicv_lnx/icv/include -I/home/b/build/opencv/cudaimgproc_ccache_log_no_equal/3rdparty/ippicv/ippicv_lnx/iw/include -I/home/b/repos/opencv/opencv_contrib/modules/cudaimgproc/include -I/home/b/build/opencv/cudaimgproc_ccache_log_no_equal/modules/cudaimgproc -I/home/b/repos/opencv/opencv_contrib/modules/cudev/include -I/home/b/repos/opencv/opencv/modules/core/include -I/home/b/repos/opencv/opencv/modules/imgproc/include -isystem=/home/b/build/opencv/cudaimgproc_ccache_log_no_equal -isystem=/usr/local/cuda/include -O3 -DNDEBUG --generate-code=arch=compute_86,code=[sm_86] -Xcompiler=-fPIC -D_FORCE_INLINES -Xcompiler=-DCVAPI_EXPORTS -Xcompiler=-fPIC "-Xcompiler=   -fsigned-char -W -Wall -Wreturn-type -Wnon-virtual-dtor -Waddress -Wsequence-point -Wformat -Wformat-security -Winit-self -Wpointer-arith  -Wuninitialized   -Wno-comment -Wno-strict-overflow -fdiagnostics-show-option -Wno-long-long -pthread -fomit-frame-pointer -ffunction-sections -fdata-sections  -msse -msse2 -msse3 -fvisibility=hidden  -Wno-undef -Wno-missing-declarations -Wno-shadow -Wno-unused-parameter -Wno-unused-but-set-variable    -O3 -DNDEBUG  -DNDEBUG" -std=c++14 -MD -MT modules/cudaimgproc/CMakeFiles/opencv_cudaimgproc.dir/src/cuda/mean_shift.cu.o -MF CMakeFiles/opencv_cudaimgproc.dir/src/cuda/mean_shift.cu.o.d -x cu -c /home/b/repos/opencv/opencv_contrib/modules/cudaimgproc/src/cuda/mean_shift.cu -o CMakeFiles/opencv_cudaimgproc.dir/src/cuda/mean_shift.cu.o
> [2023-08-28T21:55:15.968287 14016] Hostname: DESKTOP-15JFLGE
> [2023-08-28T21:55:15.968289 14016] Working directory: /home/b/build/opencv/cudaimgproc_ccache_log_no_equal/modules/cudaimgproc
> [2023-08-28T21:55:15.968297 14016] Compiler: /usr/local/cuda/bin/nvcc
> [2023-08-28T21:55:15.968299 14016] Compiler type: nvcc
> [2023-08-28T21:55:15.968366 14016] Source file: /home/b/repos/opencv/opencv_contrib/modules/cudaimgproc/src/cuda/mean_shift.cu
> [2023-08-28T21:55:15.968368 14016] Dependency file: CMakeFiles/opencv_cudaimgproc.dir/src/cuda/mean_shift.cu.o.d
> [2023-08-28T21:55:15.968369 14016] Object file: CMakeFiles/opencv_cudaimgproc.dir/src/cuda/mean_shift.cu.o
> [2023-08-28T21:55:15.968486 14016] Trying direct lookup
> [2023-08-28T21:55:15.968609 14016] Manifest key: 0a306ovodvo0kmk1ai6mr0su7qe9irr4q
> [2023-08-28T21:55:15.968637 14016] No 0a306ovodvo0kmk1ai6mr0su7qe9irr4q in local storage
> [2023-08-28T21:55:15.968721 14016] Running preprocessor
> [2023-08-28T21:55:15.968739 14016] Executing /usr/local/cuda/bin/nvcc -forward-unknown-to-host-compiler -O3 --generate-code=arch=compute_86,code=[sm_86] -Xcompiler=-fPIC -Xcompiler=-DCVAPI_EXPORTS -Xcompiler=-fPIC "-Xcompiler=   -fsigned-char -W -Wall -Wreturn-type -Wnon-virtual-dtor -Waddress -Wsequence-point -Wformat -Wformat-security -Winit-self -Wpointer-arith  -Wuninitialized   -Wno-comment -Wno-strict-overflow -fdiagnostics-show-option -Wno-long-long -pthread -fomit-frame-pointer -ffunction-sections -fdata-sections  -msse -msse2 -msse3 -fvisibility=hidden  -Wno-undef -Wno-missing-declarations -Wno-shadow -Wno-unused-parameter -Wno-unused-but-set-variable    -O3 -DNDEBUG  -DNDEBUG" -std=c++14 -DCVAPI_EXPORTS -D_USE_MATH_DEFINES -D__OPENCV_BUILD=1 -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/home/b/build/opencv/cudaimgproc_ccache_log_no_equal/3rdparty/ippicv/ippicv_lnx/icv/include -I/home/b/build/opencv/cudaimgproc_ccache_log_no_equal/3rdparty/ippicv/ippicv_lnx/iw/include -I/home/b/repos/opencv/opencv_contrib/modules/cudaimgproc/include -I/home/b/build/opencv/cudaimgproc_ccache_log_no_equal/modules/cudaimgproc -I/home/b/repos/opencv/opencv_contrib/modules/cudev/include -I/home/b/repos/opencv/opencv/modules/core/include -I/home/b/repos/opencv/opencv/modules/imgproc/include -isystem=/home/b/build/opencv/cudaimgproc_ccache_log_no_equal -isystem=/usr/local/cuda/include -DNDEBUG -D_FORCE_INLINES -x cu -E -o /mnt/wslg/runtime-dir/ccache-tmp/cpp_stdout.tmp.Jyo6x0.i /home/b/repos/opencv/opencv_contrib/modules/cudaimgproc/src/cuda/mean_shift.cu
> [2023-08-28T21:55:16.087564 14016] Got result key from preprocessor
> [2023-08-28T21:55:16.087594 14016] Result key: 3ca1pd22a6ljrqjipdto1r6janc4uhmd2
> [2023-08-28T21:55:16.087612 14016] No 3ca1pd22a6ljrqjipdto1r6janc4uhmd2 in local storage
> [2023-08-28T21:55:16.087615 14016] Running real compiler
> [2023-08-28T21:55:16.087658 14016] Executing /usr/local/cuda/bin/nvcc -forward-unknown-to-host-compiler -O3 --generate-code=arch=compute_86,code=[sm_86] -Xcompiler=-fPIC -Xcompiler=-DCVAPI_EXPORTS -Xcompiler=-fPIC "-Xcompiler=   -fsigned-char -W -Wall -Wreturn-type -Wnon-virtual-dtor -Waddress -Wsequence-point -Wformat -Wformat-security -Winit-self -Wpointer-arith  -Wuninitialized   -Wno-comment -Wno-strict-overflow -fdiagnostics-show-option -Wno-long-long -pthread -fomit-frame-pointer -ffunction-sections -fdata-sections  -msse -msse2 -msse3 -fvisibility=hidden  -Wno-undef -Wno-missing-declarations -Wno-shadow -Wno-unused-parameter -Wno-unused-but-set-variable    -O3 -DNDEBUG  -DNDEBUG" -std=c++14 -DCVAPI_EXPORTS -D_USE_MATH_DEFINES -D__OPENCV_BUILD=1 -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/home/b/build/opencv/cudaimgproc_ccache_log_no_equal/3rdparty/ippicv/ippicv_lnx/icv/include -I/home/b/build/opencv/cudaimgproc_ccache_log_no_equal/3rdparty/ippicv/ippicv_lnx/iw/include -I/home/b/repos/opencv/opencv_contrib/modules/cudaimgproc/include -I/home/b/build/opencv/cudaimgproc_ccache_log_no_equal/modules/cudaimgproc -I/home/b/repos/opencv/opencv_contrib/modules/cudev/include -I/home/b/repos/opencv/opencv/modules/core/include -I/home/b/repos/opencv/opencv/modules/imgproc/include -isystem=/home/b/build/opencv/cudaimgproc_ccache_log_no_equal -isystem=/usr/local/cuda/include -DNDEBUG -D_FORCE_INLINES -x cu -c -MD -MT modules/cudaimgproc/CMakeFiles/opencv_cudaimgproc.dir/src/cuda/mean_shift.cu.o -MF CMakeFiles/opencv_cudaimgproc.dir/src/cuda/mean_shift.cu.o.d -o CMakeFiles/opencv_cudaimgproc.dir/src/cuda/mean_shift.cu.o /home/b/repos/opencv/opencv_contrib/modules/cudaimgproc/src/cuda/mean_shift.cu
> [2023-08-28T21:55:18.092583 14016] Base dir not set, skip using relative paths
> [2023-08-28T21:55:18.092620 14016] Using default compression level 1
> [2023-08-28T21:55:18.092673 14016] Storing embedded entry #0 .o (52552 bytes) from CMakeFiles/opencv_cudaimgproc.dir/src/cuda/mean_shift.cu.o
> [2023-08-28T21:55:18.092739 14016] Storing embedded entry #1 .d (18031 bytes) from CMakeFiles/opencv_cudaimgproc.dir/src/cuda/mean_shift.cu.o.d
> [2023-08-28T21:55:18.093183 14016] Acquiring /home/b/.ccache/lock/subdir_3c.lock
> [2023-08-28T21:55:18.093227 14016] Acquired /home/b/.ccache/lock/subdir_3c.lock
> [2023-08-28T21:55:18.093232 14016] Creating /home/b/.ccache/lock/subdir_3c.alive
> [2023-08-28T21:55:18.093266 14016] Stored 3ca1pd22a6ljrqjipdto1r6janc4uhmd2 in local storage (/home/b/.ccache/3/c/a1pd22a6ljrqjipdto1r6janc4uhmd2R)
> [2023-08-28T21:55:18.093272 14016] Acquiring /home/b/.ccache/3/stats.lock
> [2023-08-28T21:55:18.093295 14016] Acquired /home/b/.ccache/3/stats.lock
> [2023-08-28T21:55:18.093298 14016] Creating /home/b/.ccache/3/stats.alive
> [2023-08-28T21:55:18.093412 14016] Releasing /home/b/.ccache/3/stats.lock
> [2023-08-28T21:55:18.093429 14016] Unlink /home/b/.ccache/3/stats.alive
> [2023-08-28T21:55:18.093441 14016] Unlink /home/b/.ccache/3/stats.lock
> [2023-08-28T21:55:18.093444 14016] Released /home/b/.ccache/3/stats.lock
> [2023-08-28T21:55:18.093447 14016] Releasing /home/b/.ccache/lock/subdir_3c.lock
> [2023-08-28T21:55:18.093459 14016] Unlink /home/b/.ccache/lock/subdir_3c.alive
> [2023-08-28T21:55:18.093470 14016] Unlink /home/b/.ccache/lock/subdir_3c.lock
> [2023-08-28T21:55:18.093473 14016] Released /home/b/.ccache/lock/subdir_3c.lock
> [2023-08-28T21:55:18.093994 14016] Added result key to manifest 0a306ovodvo0kmk1ai6mr0su7qe9irr4q
> [2023-08-28T21:55:18.094000 14016] Using default compression level 1
> [2023-08-28T21:55:18.094238 14016] Acquiring /home/b/.ccache/lock/subdir_0a.lock
> [2023-08-28T21:55:18.094260 14016] Acquired /home/b/.ccache/lock/subdir_0a.lock
> [2023-08-28T21:55:18.094264 14016] Creating /home/b/.ccache/lock/subdir_0a.alive
> [2023-08-28T21:55:18.094289 14016] Stored 0a306ovodvo0kmk1ai6mr0su7qe9irr4q in local storage (/home/b/.ccache/0/a/306ovodvo0kmk1ai6mr0su7qe9irr4qM)
> [2023-08-28T21:55:18.094295 14016] Acquiring /home/b/.ccache/0/stats.lock
> [2023-08-28T21:55:18.094315 14016] Acquired /home/b/.ccache/0/stats.lock
> [2023-08-28T21:55:18.094318 14016] Creating /home/b/.ccache/0/stats.alive
> [2023-08-28T21:55:18.094410 14016] Releasing /home/b/.ccache/0/stats.lock
> [2023-08-28T21:55:18.094424 14016] Unlink /home/b/.ccache/0/stats.alive
> [2023-08-28T21:55:18.094436 14016] Unlink /home/b/.ccache/0/stats.lock
> [2023-08-28T21:55:18.094439 14016] Released /home/b/.ccache/0/stats.lock
> [2023-08-28T21:55:18.094442 14016] Releasing /home/b/.ccache/lock/subdir_0a.lock
> [2023-08-28T21:55:18.094453 14016] Unlink /home/b/.ccache/lock/subdir_0a.alive
> [2023-08-28T21:55:18.094465 14016] Unlink /home/b/.ccache/lock/subdir_0a.lock
> [2023-08-28T21:55:18.094468 14016] Released /home/b/.ccache/lock/subdir_0a.lock
> [2023-08-28T21:55:18.094482 14016] Result: cache_miss
> [2023-08-28T21:55:18.094488 14016] Result: direct_cache_miss
> [2023-08-28T21:55:18.094490 14016] Result: local_storage_miss
> [2023-08-28T21:55:18.094492 14016] Result: local_storage_read_miss
> [2023-08-28T21:55:18.094495 14016] Result: local_storage_read_miss
> [2023-08-28T21:55:18.094497 14016] Result: local_storage_write
> [2023-08-28T21:55:18.094499 14016] Result: local_storage_write
> [2023-08-28T21:55:18.094502 14016] Result: preprocessed_cache_miss
> [2023-08-28T21:55:18.094505 14016] Acquiring /home/b/.ccache/c/0/stats.lock
> [2023-08-28T21:55:18.094599 14016] Acquired /home/b/.ccache/c/0/stats.lock
> [2023-08-28T21:55:18.094608 14016] Creating /home/b/.ccache/c/0/stats.alive
> [2023-08-28T21:55:18.094698 14016] Releasing /home/b/.ccache/c/0/stats.lock
> [2023-08-28T21:55:18.094711 14016] Unlink /home/b/.ccache/c/0/stats.alive
> [2023-08-28T21:55:18.094719 14016] Unlink /home/b/.ccache/c/0/stats.lock
> [2023-08-28T21:55:18.094720 14016] Released /home/b/.ccache/c/0/stats.lock
> [2023-08-28T21:55:18.094724 14016] Trying to acquire /home/b/.ccache/lock/auto_cleanup.lock
> [2023-08-28T21:55:18.094739 14016] Acquired /home/b/.ccache/lock/auto_cleanup.lock
> [2023-08-28T21:55:18.094741 14016] Creating /home/b/.ccache/lock/auto_cleanup.alive
> [2023-08-28T21:55:18.094807 14016] No automatic cleanup needed (size 26.8 MB, files 1048, max size 5.0 GB)
> [2023-08-28T21:55:18.094809 14016] Releasing /home/b/.ccache/lock/auto_cleanup.lock
> [2023-08-28T21:55:18.094817 14016] Unlink /home/b/.ccache/lock/auto_cleanup.alive
> [2023-08-28T21:55:18.094825 14016] Unlink /home/b/.ccache/lock/auto_cleanup.lock
> [2023-08-28T21:55:18.094826 14016] Released /home/b/.ccache/lock/auto_cleanup.lock
> [2023-08-28T21:55:18.094836 14016] Unlink /mnt/wslg/runtime-dir/ccache-tmp/stderr.tmp.E5eujZ.tmp
> [2023-08-28T21:55:18.094838 14016] Unlink /mnt/wslg/runtime-dir/ccache-tmp/stdout.tmp.rRT5VZ.tmp
> [2023-08-28T21:55:18.094840 14016] Unlink /mnt/wslg/runtime-dir/ccache-tmp/stderr.tmp.HuRcO3.tmp
> [2023-08-28T21:55:18.095081 14016] Unlink /mnt/wslg/runtime-dir/ccache-tmp/cpp_stdout.tmp.Jyo6x0.i
> [2023-08-28T21:55:18.095084 14016] Accumulated stats for inode cache: hits=0, misses=0, errors=0

maybe its the version of ccache?

Can you get ccache to work on windows for nvcc?

---

> That's the output without the changes
> 
> > --generate-code=arch=compute_86,code=[sm_86] -Xcompiler=-fPIC -D_FORCE_INLINES -Xcompiler -DCVAPI_EXPORTS -fPIC "-Xcompiler

You are right, this gets : `Result: direct_cache_hit`. Perhaps on 4.x branch (which I am using), the content of the command line is slightly different and it breaks ccache. Or perhaps it depends on the version of the CUDA Toolkit / `nvcc`.
I have tested with CUDA 10.1 and 11.4. 

> with the changes
> 
> > --generate-code=arch=compute_86,code=[sm_86] -Xcompiler=-fPIC -D_FORCE_INLINES -Xcompiler=-DCVAPI_EXPORTS -Xcompiler=-fPIC "-Xcompiler=

You also got : `Result: direct_cache_hit`
 
> Can you get ccache to work on windows for nvcc?

I don't know. Just using it on Ubuntu for the moment.